Library/OcDeviceMiscLib/SetResizableBar.c

```diff
@@ -388,3 +689,178 @@ ResizeGpuBars (
 
   return EFI_NOT_FOUND;
 }
+
+EFI_STATUS
+PciGetNextBusRange (
+  IN OUT EFI_ACPI_ADDRESS_SPACE_DESCRIPTOR  **Descriptors,
+  OUT UINT16                                *MinBus,
+  OUT UINT16                                *MaxBus,
+  OUT BOOLEAN                               *IsEnd
+  )
+{
+  *IsEnd = FALSE;
+
+  //
+  // When *Descriptors is NULL, Configuration() is not implemented, so assume
+  // range is 0~PCI_MAX_BUS
+  //
+  if ((*Descriptors) == NULL) {
+    *MinBus = 0;
+    *MaxBus = PCI_MAX_BUS;
+    return EFI_SUCCESS;
+  }
+
+  //
+  // *Descriptors points to one or more address space descriptors, which
+  // ends with a end tagged descriptor. Examine each of the descriptors,
+  // if a bus typed one is found and its bus range covers bus, this handle
+  // is the handle we are looking for.
+  //
+
+  while ((*Descriptors)->Desc != ACPI_END_TAG_DESCRIPTOR) {
+    if ((*Descriptors)->ResType == ACPI_ADDRESS_SPACE_TYPE_BUS) {
+      *MinBus = (UINT16)(*Descriptors)->AddrRangeMin;
+      *MaxBus = (UINT16)(*Descriptors)->AddrRangeMax;
+      (*Descriptors)++;
+      return (EFI_SUCCESS);
+    }
+
+    (*Descriptors)++;
+  }
+
+  if ((*Descriptors)->Desc == ACPI_END_TAG_DESCRIPTOR) {
+    *IsEnd = TRUE;
+  }
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+ResizeGpuBarsRbIo (
+  IN PCI_BAR_SIZE  Size,
+  IN BOOLEAN       Increase
+  )
+{
+  EFI_STATUS                         Status;
+  UINTN                              HandleCount;
+  EFI_HANDLE                         *HandleBuffer;
+  UINTN                              Index;
+  EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL    *PciRootBridgeIo;
+  EFI_ACPI_ADDRESS_SPACE_DESCRIPTOR  *Descriptors;
+  UINT8                              HdrType;
+  UINT8                              Bus, Dev, Func;
+  UINT16                             MinBus, MaxBus;
+  BOOLEAN                            IsEnd, HasSuccess;
+  PCI_CLASSCODE                      ClassCode;
+  UINTN                              PciAddress;
+
+  ASSERT (Size < PciBarTotal);
+
+  Status = gBS->LocateHandleBuffer (
+                  ByProtocol,
+                  &gEfiPciRootBridgeIoProtocolGuid,
+                  NULL,
+                  &HandleCount,
+                  &HandleBuffer
+                  );
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCDM: No PCI devices for RBAR support - %r\n", Status));
+    return Status;
+  }
+
+  for (Index = 0; Index < HandleCount; ++Index) {
+    Status = gBS->HandleProtocol (
+                    HandleBuffer[Index],
+                    &gEfiPciRootBridgeIoProtocolGuid,
+                    (VOID **)&PciRootBridgeIo
+                    );
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    PciRootBridgeIo->Configuration (PciRootBridgeIo, (VOID **)&Descriptors);
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
+
+    // Not sure if multiple root bridge systems even exist but this should support them
```

Comment:
They do, yeah, e.g. dual socket server stuff. Please change the comments to match the EDK II style though.